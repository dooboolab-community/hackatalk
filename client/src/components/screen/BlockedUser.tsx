import type {
  BlockedUsersQuery,
  BlockedUsersQueryResponse,
  BlockedUsersQueryVariables,
} from '../../__generated__/BlockedUsersQuery.graphql';
import { FlatList, View } from 'react-native';
import React, { FC, Suspense } from 'react';
import { graphql, useLazyLoadQuery } from 'react-relay/hooks';

import EmptyListItem from '../shared/EmptyListItem';
import { LoadingIndicator } from 'dooboo-ui';
import { RootStackNavigationProps } from '../navigation/RootStackNavigator';
import { User } from '../../types/graphql';
import UserListItem from '../shared/UserListItem';
import { getString } from '../../../STRINGS';
import styled from 'styled-components/native';
import { useProfileContext } from '../../providers/ProfileModalProvider';

const Container = styled.View`
  flex: 1;
  background-color: transparent;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

interface Props {
  navigation: RootStackNavigationProps<'default'>;
}

const blockedUsersQuery = graphql`
  query BlockedUsersQuery {
    blockedUsers {
      id
      email
      name
      nickname
      hasBlocked
      photoURL
      thumbURL
      photoURL
      statusMessage
    }
  }
`;

const ContentContainer: FC = () => {
  const { showModal } = useProfileContext();

  const {
    blockedUsers = [],
  }: BlockedUsersQueryResponse = useLazyLoadQuery<BlockedUsersQuery>(
    blockedUsersQuery,
    {},
    { fetchPolicy: 'store-or-network' },
  );

  const renderItem = ({
    item,
    index,
  }: {
    item: User;
    index: number;
  }): React.ReactElement => {
    const itemTestID = `user-list-item${index}`;

    const pressUserItem = (): void => {
      showModal({
        user: item,
        isFriend: false,
        hideButtons: true,
      });
    };

    return (
      <UserListItem testID={itemTestID} user={item} onPress={pressUserItem} />
    );
  };

  return (
    <FlatList
      style={{ alignSelf: 'stretch' }}
      contentContainerStyle={
        blockedUsers?.length === 0
          ? {
              flex: 1,
              alignSelf: 'stretch',
              alignItems: 'center',
              justifyContent: 'center',
            }
          : null
      }
      keyExtractor={(_, index): string => index.toString()}
      // @ts-ignore
      data={blockedUsers}
      renderItem={renderItem}
      ListEmptyComponent={
        <EmptyListItem>{getString('NO_BANNED_USER')}</EmptyListItem>
      }
      ListFooterComponent={<View style={{ height: 60 }} />}
      onEndReachedThreshold={0.1}
    />
  );
};

const Page: FC<Props> = () => {
  return (
    <Container>
      <Suspense fallback={<LoadingIndicator />}>
        <ContentContainer />
      </Suspense>
    </Container>
  );
};

export default Page;
