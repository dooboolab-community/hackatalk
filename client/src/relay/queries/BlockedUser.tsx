import {graphql} from 'react-relay';

export const createBlockedUserMutation = graphql`
  mutation BlockedUserCreateMutation($blockedUserId: String!) {
    createBlockedUser(blockedUserId: $blockedUserId) {
      blockedUser {
        id
        email
        name
        nickname
        hasBlocked
      }
    }
  }
`;

export const deleteBlockedUserMutation = graphql`
  mutation BlockedUserDeleteMutation($blockedUserId: String!) {
    deleteBlockedUser(blockedUserId: $blockedUserId) {
      blockedUser {
        id
        email
        name
        nickname
        hasBlocked
      }
    }
  }
`;
