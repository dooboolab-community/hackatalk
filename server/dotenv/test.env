JWT_SECRET=dooboolab
DATABASE_URL="postgresql://test:test!@localhost:5432/postgres?schema=test"
