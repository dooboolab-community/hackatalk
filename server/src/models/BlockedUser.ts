import { objectType } from '@nexus/schema';

export const BlockedUser = objectType({
  name: 'BlockedUser',
  definition(t) {
    t.model.createdAt();
    t.model.updatedAt();
    t.model.deletedAt();
    t.field('user', { type: 'User', nullable: true });
    t.field('blockedUser', { type: 'User', nullable: true });
  },
});
