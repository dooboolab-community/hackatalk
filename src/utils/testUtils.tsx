import React, { ReactElement } from 'react';

import RootProvider from '../providers';
import { ThemeType } from '../types';

export const createTestElement = (
  child: ReactElement,
  themeType?: ThemeType,
): ReactElement => (
  <RootProvider initialThemeType={themeType}>{child}</RootProvider>
);

export const createTestProps = (
  obj: object = {},
  moreScreenProps?: object,
): object | unknown | any => ({
  navigation: {
    navigate: jest.fn(),
    goBack: jest.fn(),
    resetRoot: jest.fn(),
  },
  screenProps: {
    changeThemeType: jest.fn(),
    ...moreScreenProps,
  },
  ...obj,
});
