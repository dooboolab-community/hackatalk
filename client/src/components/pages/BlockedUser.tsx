import {FlatList, ListRenderItem, View} from 'react-native';
import React, {FC, Suspense, useMemo} from 'react';

import EmptyListItem from '../uis/EmptyListItem';
import {LoadingIndicator} from 'dooboo-ui';
import type {UserBlockedUsersQuery} from '../../__generated__/UserBlockedUsersQuery.graphql';
import UserListItem from '../uis/UserListItem';
import {blockedUsersQuery} from '../../relay/queries/User';
import {getString} from '../../../STRINGS';
import styled from '@emotion/native';
import {useLazyLoadQuery} from 'react-relay';
import {useProfileContext} from '../../providers/ProfileModalProvider';

const Container = styled.View`
  flex: 1;
  background-color: transparent;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

const ContentContainer: FC = () => {
  const {showModal} = useProfileContext();

  const response = useLazyLoadQuery<UserBlockedUsersQuery>(
    blockedUsersQuery,
    {},
    {fetchPolicy: 'store-or-network'},
  );

  const blockedUsers = useMemo(() => {
    return (
      response.blockedUsers?.filter(
        (x): x is NonNullable<typeof x> => x !== null,
      ) ?? []
    );
  }, [response]);

  const renderItem: ListRenderItem<typeof blockedUsers[number]> = ({item}) => {
    const pressUserItem = (): void => {
      showModal({
        user: item,
        hideButtons: true,
      });
    };

    return <UserListItem user={item} onPress={pressUserItem} />;
  };

  return (
    <FlatList
      style={{alignSelf: 'stretch'}}
      contentContainerStyle={
        blockedUsers?.length === 0
          ? {
              flex: 1,
              alignSelf: 'stretch',
              alignItems: 'center',
              justifyContent: 'center',
            }
          : null
      }
      keyExtractor={(_, index): string => index.toString()}
      data={blockedUsers}
      renderItem={renderItem}
      ListEmptyComponent={
        <EmptyListItem>{getString('NO_BANNED_USER')}</EmptyListItem>
      }
      ListFooterComponent={<View style={{height: 60}} />}
      onEndReachedThreshold={0.1}
    />
  );
};

const Page: FC = () => {
  return (
    <Container>
      <Suspense fallback={<LoadingIndicator />}>
        <ContentContainer />
      </Suspense>
    </Container>
  );
};

export default Page;
