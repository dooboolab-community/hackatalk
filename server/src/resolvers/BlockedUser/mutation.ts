import { mutationField, nonNull, stringArg } from '@nexus/schema';

import { getUserId } from '../../utils/auth';

export const createBlockedUser = mutationField('createBlockedUser', {
  type: 'BlockedUser',

  args: {
    blockedUserId: nonNull(stringArg()),
  },

  resolve: (_, { blockedUserId }, ctx) => {
    const userId = getUserId(ctx);

    return ctx.prisma.blockedUser.create({
      data: {
        user: {
          connect: {
            id: userId,
          },
        },
        blockedUser: {
          connect: {
            id: blockedUserId,
          },
        },
      },
      include: {
        user: true,
        blockedUser: true,
      },
    });
  },
});

export const deleteBlockedUser = mutationField('deleteBlockedUser', {
  type: 'BlockedUser',

  args: {
    blockedUserId: nonNull(stringArg()),
  },

  resolve: (_, { blockedUserId }, ctx) => {
    const userId = getUserId(ctx);

    return ctx.prisma.blockedUser.delete({
      where: {
        userId_blockedUserId: {
          userId,
          blockedUserId,
        },
      },
      include: {
        user: true,
        blockedUser: true,
      },
    });
  },
});
