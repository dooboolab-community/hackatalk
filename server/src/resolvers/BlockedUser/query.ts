import { queryField, stringArg } from '@nexus/schema';

import { getUserId } from '../../utils/auth';

export const blockedUsers = queryField('blockedUsers', {
  type: 'BlockedUser',
  list: true,
  nullable: true,

  description: 'Arguments are not needed. Only find blocked users of signed in user',

  resolve: (_, __, ctx) => {
    const userId = getUserId(ctx);

    return ctx.prisma.blockedUser.findMany({
      where: {
        userId,
      },
    });
  },
});
