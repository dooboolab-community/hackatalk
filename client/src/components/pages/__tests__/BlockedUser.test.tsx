import 'react-native';

import {MockPayloadGenerator, createMockEnvironment} from 'relay-test-utils';

import BlockedUser from '../BlockedUser';
import React from 'react';
import {createTestElement} from '../../../../test/testUtils';
import {render} from '@testing-library/react-native';

const mockEnvironment = createMockEnvironment();

mockEnvironment.mock.queueOperationResolver((operation) =>
  MockPayloadGenerator.generate(operation),
);

describe('Rendering', () => {
  it('renders without crashing', async () => {
    const component = createTestElement(<BlockedUser />, {
      environment: mockEnvironment,
    });

    const screen = render(component);
    const json = screen.toJSON();

    expect(json).toMatchSnapshot();
    expect(json).toBeTruthy();
  });
});
