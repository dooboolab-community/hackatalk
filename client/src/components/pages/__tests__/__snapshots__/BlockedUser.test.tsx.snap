// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering renders without crashing 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <RCTScrollView
      ListEmptyComponent={
        <Shared
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "flex": 1,
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          No banned user
        </Shared>
      }
      ListFooterComponent={
        <View
          style={
            Object {
              "height": 60,
            }
          }
        />
      }
      contentContainerStyle={null}
      data={
        Array [
          Object {
            "__fragmentOwner": Object {
              "cacheConfig": Object {
                "force": true,
              },
              "identifier": "7bd15cc4e160652eb15d970256188c5b{}",
              "node": Object {
                "fragment": Object {
                  "abstractKey": null,
                  "argumentDefinitions": Array [],
                  "kind": "Fragment",
                  "metadata": null,
                  "name": "UserBlockedUsersQuery",
                  "selections": Array [
                    Object {
                      "alias": null,
                      "args": null,
                      "concreteType": "User",
                      "kind": "LinkedField",
                      "name": "blockedUsers",
                      "plural": true,
                      "selections": Array [
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "id",
                          "storageKey": null,
                        },
                        Object {
                          "args": null,
                          "kind": "FragmentSpread",
                          "name": "ProfileModal_user",
                        },
                        Object {
                          "args": null,
                          "kind": "FragmentSpread",
                          "name": "UserListItem_user",
                        },
                      ],
                      "storageKey": null,
                    },
                  ],
                  "type": "Query",
                },
                "hash": "de6f1c0db8178def0afba2a4c90484fc",
                "kind": "Request",
                "operation": Object {
                  "argumentDefinitions": Array [],
                  "kind": "Operation",
                  "name": "UserBlockedUsersQuery",
                  "selections": Array [
                    Object {
                      "alias": null,
                      "args": null,
                      "concreteType": "User",
                      "kind": "LinkedField",
                      "name": "blockedUsers",
                      "plural": true,
                      "selections": Array [
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "id",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "photoURL",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "name",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "nickname",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "hasBlocked",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "statusMessage",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "isFriend",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "isOnline",
                          "storageKey": null,
                        },
                      ],
                      "storageKey": null,
                    },
                  ],
                },
                "params": Object {
                  "cacheID": "7bd15cc4e160652eb15d970256188c5b",
                  "id": null,
                  "metadata": Object {},
                  "name": "UserBlockedUsersQuery",
                  "operationKind": "query",
                  "text": "query UserBlockedUsersQuery {
  blockedUsers {
    id
    ...ProfileModal_user
    ...UserListItem_user
  }
}

fragment ProfileModal_user on User {
  id
  photoURL
  name
  nickname
  hasBlocked
  statusMessage
  isFriend
}

fragment UserListItem_user on User {
  id
  photoURL
  nickname
  name
  statusMessage
  isOnline
  hasBlocked
}
",
                },
              },
              "variables": Object {},
            },
            "__fragments": Object {
              "ProfileModal_user": Object {},
              "UserListItem_user": Object {},
            },
            "__id": "<User-mock-id-1>",
            "__isWithinUnmatchedTypeRefinement": false,
            "id": "<User-mock-id-1>",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={0.1}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "alignSelf": "stretch",
        }
      }
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "width": "100%",
                },
                undefined,
              ]
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="peer-button"
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#151C22",
                      "borderBottomColor": "#000",
                      "borderBottomWidth": 1,
                      "borderLeftColor": "#000",
                      "borderRightColor": "#000",
                      "borderTopColor": "#000",
                      "flexDirection": "row",
                      "height": 80,
                      "justifyContent": "flex-start",
                      "paddingBottom": 0,
                      "paddingLeft": 20,
                      "paddingRight": 20,
                      "paddingTop": 0,
                    },
                    undefined,
                  ]
                }
                testID="userListItem-wrapper"
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "height": 40,
                        "justifyContent": "center",
                        "width": 40,
                      },
                      undefined,
                    ]
                  }
                >
                  <Image
                    resizeMode="cover"
                    source={
                      Object {
                        "uri": "<mock-value-for-field-\\"photoURL\\">",
                      }
                    }
                    style={
                      Array [
                        Object {
                          "borderBottomLeftRadius": 20,
                          "borderBottomRightRadius": 20,
                          "borderTopLeftRadius": 20,
                          "borderTopRightRadius": 20,
                          "height": 40,
                          "width": 40,
                        },
                        undefined,
                      ]
                    }
                  />
                  <View
                    style={
                      Object {
                        "left": 0,
                        "position": "absolute",
                        "top": 0,
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "white",
                        "fontSize": 14,
                        "fontWeight": "bold",
                        "marginLeft": 12,
                        "width": 200,
                      },
                      undefined,
                    ]
                  }
                >
                  &lt;mock-value-for-field-"nickname"&gt;
                </Text>
                <Text
                  ellipsizeMode="tail"
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "white",
                        "fontSize": 12,
                        "maxWidth": 120,
                        "overflow": "hidden",
                        "paddingBottom": 4,
                        "paddingLeft": 8,
                        "paddingRight": 8,
                        "paddingTop": 4,
                        "position": "absolute",
                        "right": 20,
                      },
                      undefined,
                    ]
                  }
                >
                  &lt;mock-value-for-field-"statusMessage"&gt;
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Object {
                "height": 60,
              }
            }
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</RNCSafeAreaProvider>
`;
